diff --git a/relation.rules.inc b/relation.rules.inc
index ed5f8a3..8c13800 100644
--- a/relation.rules.inc
+++ b/relation.rules.inc
@@ -100,7 +100,12 @@ function relation_rules_get_type_options() {
  * Callback for creating a relation, in the form Rules wants it.
  */
 function relation_rules_create($values = array()) {
-  return relation_create($values['relation_type'], array());
+  $relation = relation_create($values['relation_type'], array());
+  foreach ($values as $key => $value) {
+    $relation->{$key} = $value;
+  }
+
+  return $relation;
 }
 
 /**