--- cors.module.original	2013-04-01 12:37:39.712383051 -0700
+++ cors.module	2013-04-01 12:37:18.050396658 -0700
@@ -46,6 +46,36 @@
     '#rows' => 10,
   );
 
+  // Methods: http://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html
+  $methods_options = array('OPTIONS' => 'OPTIONS', 'GET' => 'GET', 'HEAD' => 'HEAD', 'POST' => 'POST', 'PUT' => 'PUT', 'DELETE' => 'DELETE', 'TRACE' => 'TRACE', 'CONNECT' => 'CONNECT');
+  $methods_default_value = variable_get('cors_methods', array(0, 1, 3, 4));
+  $form['cors_methods'] = array(
+    '#type' => 'checkboxes',
+    '#title' => t('Methods'),
+    '#options' => $methods_options,
+    '#default_value' => $methods_default_value,
+  );
+
+  // Headers: http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html
+  $headers_default_value_multiline = html_entity_decode(implode('&#13;&#10;', array('Origin', 'Content-Type', 'Accept', 'Authorization')));
+  $headers_default_value = variable_get('cors_headers', $headers_default_value_multiline);
+  $form['cors_headers'] = array(
+    '#type' => 'textarea',
+    '#title' => t('Headers'),
+    '#description' => t('A list of headers to enable for CORS. Enter one value per line.'),
+    '#default_value' => $headers_default_value,
+    '#rows' => 6,
+  );
+
+  // Allow Credentials: http://www.w3.org/TR/access-control/#access-control-allow-credentials-response-header
+  $allow_credentials_default_value = variable_get('cors_allow_credentials', 0);
+  $form['cors_allow_credentials'] = array(
+    '#type' => 'checkbox',
+    '#title' => t('Set Access-Control-Allow-Credentials = true'),
+    '#description' => t('Required for AJAX "withCredentials" requests'),
+    '#default_value' => $allow_credentials_default_value,
+  );
+
   $form['submit'] = array(
     '#type' => 'submit',
     '#value' => t('Save configuration'),
@@ -69,6 +99,10 @@
   }
 
   variable_set('cors_domains', $settings);
+  variable_set('cors_methods', $form_state['values']['cors_methods']);
+  variable_set('cors_headers', $form_state['values']['cors_headers']);
+  variable_set('cors_allow_credentials', $form_state['values']['cors_allow_credentials']);
+  drupal_set_message(t('Settings saved successfully.'));
 }
 
 /**
@@ -88,8 +122,28 @@
     }
   }
   $allowed_domains = trim($allowed_domains);
+
+  // methods
+  $methods = variable_get('cors_methods', NULL);
+  $allowed_methods = '';
+  if (isset($methods) && is_array($methods)) {
+    foreach ($methods as $method) {
+      if (!empty($method)) $allowed_methods .= $method . ', ';
+    }
+    $allowed_methods = trim($allowed_methods, ', ');
+  }
+
+  // headers
+  $allowed_headers = str_replace(html_entity_decode('&#13;&#10;'), ', ', variable_get('cors_headers', NULL));
+
+  // allowed credentials
+  $allow_credentials = variable_get('cors_allow_credentials', NULL);
+
   if (strlen($allowed_domains) > 0) {
     drupal_add_http_header('Access-Control-Allow-Origin', $allowed_domains, TRUE);
+    if (!empty($allowed_methods)) drupal_add_http_header('Access-Control-Allow-Methods', $allowed_methods, TRUE);
+    if (!empty($allowed_headers)) drupal_add_http_header('Access-Control-Allow-Headers', $allowed_headers, TRUE);
+    if (!empty($allow_credentials)) drupal_add_http_header('Access-Control-Allow-Credentials', 'true', TRUE);
   }
 }
 
